### DISCUSSION TITLE
State of Final Cut API

### DISCUSSION THREAD

**Discussion by pravdomil** [CommandPost/CommandPost#3231] [General]
Simple task: get selected clips file paths.

Premiere: `app.project.activeSequence.getSelection().map(a => a.projectItem.getMediaPath())`

Final Cut: let me read what is shown in UI, well I cannot tell the paths

Is is only because Workflow Extensions give us 7 functions to work with?
```
func ProExtensionHostSingleton() -> NSObjectProtocol?
func add(FCPXTimelineObserver)
func movePlayhead(to: CMTime) -> CMTime
func playheadTime() -> CMTime
func remove(FCPXTimelineObserver)
func activeSequenceChanged()
func playheadTimeChanged()
```

**Comment by latenitefilms** [comment]
Final Cut Pro doesn't have **any** official API apart from [FxPlug4](https://developer.apple.com/documentation/professional_video_applications/fxplug?language=objc) for effects, and [Workflow Extensions](https://developer.apple.com/documentation/professional_video_applications/workflow_extensions?language=objc) to allow you to integrate your app's workflow within the Final Cut Pro interface to streamline data exchange.

CommandPost uses a wide range of methods such as the Accessibility API, reading files at the file system level, and controlling the playhead position using a Workflow Extension to build our own API for controlling Final Cut Pro.

The CommandPost Lua API is also still always an evolving work-in-progress, that @randomeizer and myself update when we get a chance, and generally speaking only when we need it to do something in a CommandPost Plugin that it doesn't currently do.

Final Cut Pro doesn't always play nicely in terms of the Accessibility API, and a lot of the time they break things between versions, so we need to find workarounds to things. We also try to allow CommandPost to work in all languages that Final Cut Pro supports, but something Apple just weird things in terms of the Accessibility API in languages other than English, so again, we need to do workarounds for that too.

You can use [UI Browser](https://latenightsw.com/freeware/ui-browser/) or Xcode's Accessibility Inspector to explore the Accessibility Tree and Objects in Final Cut Pro to determine what can and cannot be accessed.
